lib/facilities/local_store.rb

```diff
@@ -0,0 +1,109 @@
+# frozen_string_literal: true
+require 'facilities/bulk_client'
+
+module Facilities
+  class LocalStore
+    attr_reader :adapter
+
+    GIS_CHECK_FREQUENCY = 1.hour
+    BATCH_SIZE = 1000.0
+    OPEN_TIMEOUT = 2
+    REQUEST_TIMEOUT = 6
+    Pointer = Struct.new(:x, :y, :facility)
+
+    def initialize(url, _type, adapter)
+      @url = url
+      @adapter = adapter
+      @mutex = Mutex.new
+      @last_check = Time.new(0).utc
+      @current_gis_update = 0
+      @client = BulkClient.new(@url)
+      @index = {}
+      @coordinates = []
+      # check_for_freshness
+    end
+
+    def get(id)
+      check_for_freshness
+      @mutex.synchronize do
+        return @index[id]
+      end
+    end
+
+    def query(bbox, services = nil)
+      check_for_freshness
+      filter = services ? @adapter.with_services(services) : ->(_) { true }
+      @mutex.synchronize do
+        @coordinates.select(&(within bbox)).sort_by(&(dist_from_center bbox)).map(&:facility).select(&filter)
+      end
+    end
+
+    def within(bbox)
+      x_min, x_max = min_max(bbox[0], bbox[2])
+      y_min, y_max = min_max(bbox[1], bbox[3])
+      lambda do |point|
```

Comment:
I think this is technically correct but not great in terms of encapsulation. It's a little weird to have the `within` method be on the LocalStore class, since it's not a question you're asking of that object, it's really a question for the coordinates themselves. Why not have Pointer be a proper class and then implement these sorts of methods on it? Then the select above could be implemented as:

`@coordinates.select { |c| c.within?(bbox) }`

I think that reads a lot more easily and is more Rubyish.